File: zig/native_test.zig
const std = @import("std");
const testing = std.testing;
const skg = @import("root.zig");
const native = skg.native;

test "native structs maps lists defaults enums and nullable records" {
    const Record = struct { id: u16, note: ?[]const u8 };
    const Config = struct {
        name: []const u8,
        enabled: bool = true,
        mode: enum { local, remote } = .local,
        headers: std.StringHashMap([]const u8),
        records: []?Record,
        matrix: [2][]const i32,
        pub const skg_fields = .{ .name = "service.name" };
    };
    var result = skg.decodeSource(Config, testing.allocator,
        \\"service.name": "demo"
        \\mode: "remote"
        \\headers { "Content-Type": "json" }
        \\records [null { id: 3 note: null }]
        \\matrix: [[1, 2], [3]]
        \\unknown { ignored: true }
    , "config.skg", .{});
    defer result.deinit();
    const value = result.value orelse return error.DecodeFailed;
    try testing.expectEqualStrings("demo", value.name);
    try testing.expect(value.enabled);
    try testing.expectEqual(.remote, value.mode);
    try testing.expectEqualStrings("json", value.headers.get("Content-Type").?);
    try testing.expect(value.records[0] == null);
    try testing.expectEqual(@as(u16, 3), value.records[1].?.id);
    try testing.expect(value.records[1].?.note == null);
    try testing.expectEqualSlices(i32, &.{ 1, 2 }, value.matrix[0]);
}

test "native managed map allocator survives result movement and source release" {
    const Config = struct { map: std.StringHashMap(i64) };
    const src = try testing.allocator.dupe(u8, "map { a: 1 }");
    var result = skg.decodeSource(Config, testing.allocator, src, "map.skg", .{});
    testing.allocator.free(src);
    defer result.deinit();
    try testing.expect(result.value != null);
    for (0..100) |i| {
        const key = try std.fmt.allocPrint(result.arena.?.allocator(), "key{d}", .{i});
        try result.value.?.map.put(key, @intCast(i));
    }
    try testing.expectEqual(@as(i64, 99), result.value.?.map.get("key99").?);
}

test "native defaults are copied optional absent and ignored native fields" {
    const Config = struct {
        name: []const u8 = "default",
        optional: ?i32,
        local_only: u8 = 7,
        pub const skg_fields = .{ .local_only = null };
    };
    var result = skg.decodeSource(Config, testing.allocator, "local_only: 9", "defaults.skg", .{});
    defer result.deinit();
    const value = result.value orelse return error.DecodeFailed;
    try testing.expectEqualStrings("default", value.name);
    try testing.expect(value.optional == null);
    try testing.expectEqual(@as(u8, 7), value.local_only);
}

test "native errors include escaped field paths and source positions" {
    const Config = struct {
        values: []u8,
        pub const skg_fields = .{ .values = "a/b~c" };
    };
    var result = skg.decodeSource(Config, testing.allocator, "\"a/b~c\": [1, 256]", "range.skg", .{});
    defer result.deinit();
    try testing.expect(result.value == null);
    const d = result.diagnostic orelse return error.NoDiagnostic;
    try testing.expectEqual(native.Code.number_out_of_range, d.code);
    try testing.expectEqualStrings("/a~1b~0c/1", d.field_path);
    try testing.expectEqualStrings("range.skg", d.source.path);
    try testing.expectEqual(@as(u32, 1), d.source.line);
    try testing.expectEqual(@as(u32, 1), d.source.col);
}

test "native missing null wrong type unknown enum and strict unknowns" {
    const Required = struct { count: u8 };
    const cases = [_]struct { src: []const u8, code: native.Code }{
        .{ .src = "", .code = .missing_field },
        .{ .src = "count: null", .code = .type_mismatch },
        .{ .src = "count: -1", .code = .number_out_of_range },
        .{ .src = "count: \"3\"", .code = .type_mismatch },
    };
    for (cases) |case| {
        var result = skg.decodeSource(Required, testing.allocator, case.src, "bad.skg", .{});
        defer result.deinit();
        try testing.expect(result.value == null);
        try testing.expectEqual(case.code, result.diagnostic.?.code);
    }
    const Config = struct { mode: enum { local } };
    var bad_enum = skg.decodeSource(Config, testing.allocator, "mode: \"remote\"", "", .{});
    defer bad_enum.deinit();
    try testing.expectEqual(native.Code.invalid_enum, bad_enum.diagnostic.?.code);
    var unknown = skg.decodeSource(Required, testing.allocator, "count: 1 extra: true", "", .{ .unknown_fields = .reject });
    defer unknown.deinit();
    try testing.expectEqual(native.Code.unknown_field, unknown.diagnostic.?.code);
    try testing.expectEqualStrings("/extra", unknown.diagnostic.?.field_path);
}

test "native exact numeric conversion with explicit lossy opt in" {
    const Float32 = struct { value: f32 };
    var exact = skg.decodeSource(Float32, testing.allocator, "value: 0.5", "", .{});
    defer exact.deinit();
    try testing.expectEqual(@as(f32, 0.5), exact.value.?.value);
    var inexact = skg.decodeSource(Float32, testing.allocator, "value: 0.1", "", .{});
    defer inexact.deinit();
    try testing.expectEqual(native.Code.inexact_number, inexact.diagnostic.?.code);
    var lossy = skg.decodeSource(Float32, testing.allocator, "value: 0.1", "", .{ .allow_lossy_numbers = true });
    defer lossy.deinit();
    try testing.expectEqual(@as(f32, 0.1), lossy.value.?.value);
    const Float64 = struct { value: f64 };
    var max_int = skg.decodeSource(Float64, testing.allocator, "value: 9223372036854775807", "", .{});
    defer max_int.deinit();
    try testing.expectEqual(native.Code.inexact_number, max_int.diagnostic.?.code);
    var min_int = skg.decodeSource(Float64, testing.allocator, "value: -9223372036854775808", "", .{});
    defer min_int.deinit();
    try testing.expect(min_int.value != null);
}

test "native pointers sentinel strings unmanaged maps dynamic values" {
    const Config = struct {
        text: [:0]const u8,
        record: *const struct { id: i64 },
        map: std.StringHashMapUnmanaged(?i32),
        dynamic: skg.Value,
    };
    var result = skg.decodeSource(Config, testing.allocator, "text: \"hello\" record { id: 2 } map { a: null b: 3 } dynamic: [[{x: 1}, null]]", "", .{});
    defer result.deinit();
    const value = result.value orelse return error.DecodeFailed;
    try testing.expectEqual(@as(u8, 0), value.text[value.text.len]);
    try testing.expectEqual(@as(i64, 2), value.record.id);
    try testing.expect(value.map.get("a").? == null);
    try testing.expectEqual(@as(i32, 3), value.map.get("b").?.?);
    try testing.expect(value.dynamic == .array);
}

test "native hooks decode and validate with local diagnostics" {
    const Port = struct {
        value: u16,
        pub fn skgDecode(ctx: *native.Context, input: skg.Value) native.Error!@This() {
            return .{ .value = try ctx.decode(u16, input) };
        }
        pub fn skgValidate(self: @This(), ctx: *native.Context) native.Error!void {
            if (self.value == 0) return ctx.fail(.custom_error, "port must be nonzero");
        }
    };
    const Config = struct { port: Port };
    var good = skg.decodeSource(Config, testing.allocator, "port: 8080", "", .{});
    defer good.deinit();
    try testing.expectEqual(@as(u16, 8080), good.value.?.port.value);
    var bad = skg.decodeSource(Config, testing.allocator, "port: 0", "", .{});
    defer bad.deinit();
    try testing.expectEqualStrings("/port", bad.diagnostic.?.field_path);
    try testing.expectEqualStrings("port must be nonzero", bad.diagnostic.?.message);
}

test "native import source provenance and overlay finalization" {
    var dir = testing.tmpDir(.{});
    defer dir.cleanup();
    try dir.dir.writeFile(.{ .sub_path = "base.skg", .data = "count: 256\nremoved: 1" });
    try dir.dir.writeFile(.{ .sub_path = "main.skg", .data = "import \"base.skg\"\n@delete removed" });
    const path = try dir.dir.realpathAlloc(testing.allocator, "main.skg");
    defer testing.allocator.free(path);
    const Config = struct { count: u8 };
    var result = skg.decodeFile(Config, testing.allocator, path, .{});
    defer result.deinit();
    try testing.expectEqual(native.Code.number_out_of_range, result.diagnostic.?.code);
    try testing.expectEqualStrings("base.skg", std.fs.path.basename(result.diagnostic.?.source.path));
    const Wide = struct { count: u16 };
    var wide = skg.decodeFile(Wide, testing.allocator, path, .{ .unknown_fields = .reject });
    defer wide.deinit();
    try testing.expectEqual(@as(u16, 256), wide.value.?.count);
}

fn allocationCase(allocator: std.mem.Allocator) !void {
    const Config = struct { name: []const u8, map: std.StringHashMap([]?u16) };
    var result = skg.decodeSource(Config, allocator, "name: \"a\" map { items: [1, null, 3] }", "memory.skg", .{});
    defer result.deinit();
    if (result.value == null) {
        if (result.diagnostic) |d| {
            if (d.code == .out_of_memory) return error.OutOfMemory;
        }
        return error.UnexpectedDecodeFailure;
    }
}

test "native allocation failures clean up every owned allocation" {
    try testing.checkAllAllocationFailures(testing.allocator, allocationCase, .{});
}

test "native parse failures keep parser diagnostics and no partial result" {
    const Config = struct { value: i32 };
    var result = skg.decodeSource(Config, testing.allocator, "value: [", "broken.skg", .{});
    defer result.deinit();
    try testing.expect(result.value == null);
    const d = result.diagnostic.?;
    try testing.expectEqual(native.Code.parse_error, d.code);
    try testing.expectEqual(skg.ast.ErrorCode.UNTERMINATED_ARRAY, d.parse_diagnostic.?.code);
    try testing.expectEqualStrings("broken.skg", d.source.path);
}

test "native recursive hooks are bounded" {
    const Recursive = struct {
        pub fn skgDecode(ctx: *native.Context, value: skg.Value) native.Error!@This() {
            return ctx.decode(@This(), value);
        }
    };
    var result = skg.decodeSource(Recursive, testing.allocator, "", "", .{});
    defer result.deinit();
    try testing.expect(result.value == null);
    try testing.expectEqual(native.Code.nesting_too_deep, result.diagnostic.?.code);
}

test "native mapping errors and fixed array mismatch" {
    const Duplicate = struct {
        a: i32 = 0,
        b: i32 = 0,
        pub const skg_fields = .{ .b = "a" };
    };
    var duplicate = skg.decodeSource(Duplicate, testing.allocator, "", "", .{});
    defer duplicate.deinit();
    try testing.expectEqual(native.Code.unsupported_type, duplicate.diagnostic.?.code);
    const Typo = struct {
        a: i32 = 0,
        pub const skg_fields = .{ .typo = "name" };
    };
    var typo = skg.decodeSource(Typo, testing.allocator, "", "", .{});
    defer typo.deinit();
    try testing.expectEqual(native.Code.unsupported_type, typo.diagnostic.?.code);
    const Fixed = struct { items: [2]u8 };
    var fixed = skg.decodeSource(Fixed, testing.allocator, "items: [1]", "", .{});
    defer fixed.deinit();
    try testing.expectEqual(native.Code.type_mismatch, fixed.diagnostic.?.code);
}

test "native byte loading ignores filesystem imports" {
    const Config = struct { name: []const u8 };
    var result = skg.decodeSource(Config, testing.allocator, "import \"does-not-exist.skg\" name: \"local\"", "local.skg", .{});
    defer result.deinit();
    try testing.expectEqualStrings("local", result.value.?.name);
}

test "native float overflow remains an error with lossy conversion enabled" {
    const Config = struct { value: f32 };
    var result = skg.decodeSource(Config, testing.allocator, "value: 10000000000000000000000000000000000000000.0", "", .{ .allow_lossy_numbers = true });
    defer result.deinit();
    try testing.expect(result.value == null);
    try testing.expectEqual(native.Code.number_out_of_range, result.diagnostic.?.code);
}

test "native mutable defaults are independently owned" {
    const Config = struct { text: []const u8 = "abc", numbers: []const u8 = &.{ 1, 2 } };
    var a = skg.decodeSource(Config, testing.allocator, "", "", .{});
    defer a.deinit();
    var b = skg.decodeSource(Config, testing.allocator, "", "", .{});
    defer b.deinit();
    try testing.expect(a.value.?.text.ptr != b.value.?.text.ptr);
    try testing.expect(a.value.?.numbers.ptr != b.value.?.numbers.ptr);
}

test "native byte arrays are range checked and no failed value escapes" {
    const Config = struct { text: []const u8, later: u8 };
    var result = skg.decodeSource(Config, testing.allocator, "text: [65, 66] later: 256", "", .{});
    defer result.deinit();
    try testing.expect(result.value == null);
    try testing.expectEqualStrings("/later", result.diagnostic.?.field_path);
}

File: go/native_test.go
package skg

import (
	"errors"
	"fmt"
	"math"
	"os"
	"path/filepath"
	"reflect"
	"testing"
	"time"
	"unsafe"
)

func expectNativeError(t *testing.T, err error, code NativeCode, path string) *DecodeError {
	t.Helper()
	var d *DecodeError
	if !errors.As(err, &d) || d.Code != code || d.FieldPath != path {
		t.Fatalf("got %#v (%v), want %s at %q", d, err, code, path)
	}
	return d
}

func TestNativePresenceAndCompatibility(t *testing.T) {
	type Config struct {
		Count    uint8  `skg:"count"`
		Optional *uint8 `skg:"optional"`
		Values   []int  `skg:"values"`
	}
	for _, tc := range []struct {
		src  string
		code NativeCode
	}{
		{"", NativeMissingField}, {"count: null", NativeTypeMismatch},
		{"count: -1", NativeNumberOutOfRange}, {"count: 256", NativeNumberOutOfRange},
		{"count: \"1\"", NativeTypeMismatch},
	} {
		value, err := DecodeSource[Config]([]byte(tc.src), "config.skg", DecodeOptions{})
		expectNativeError(t, err, tc.code, "/count")
		if !reflect.DeepEqual(value, Config{}) {
			t.Fatal("failed result was not zero")
		}
	}
	value, err := DecodeSource[Config]([]byte("count: 0 ignored { any: true }"), "", DecodeOptions{})
	if err != nil || value.Count != 0 || value.Optional != nil || value.Values != nil {
		t.Fatalf("%#v %v", value, err)
	}
	_, err = DecodeSource[Config]([]byte("count: 1 extra: true"), "", DecodeOptions{RejectUnknownFields: true})
	expectNativeError(t, err, NativeUnknownField, "/extra")
	legacy := Config{Count: 7}
	if err := Unmarshal([]byte("optional: null"), &legacy); err != nil || legacy.Count != 7 {
		t.Fatalf("legacy absent: %#v %v", legacy, err)
	}
	if err := Unmarshal([]byte("count: null"), &legacy); err != nil || legacy.Count != 0 {
		t.Fatalf("legacy null: %#v %v", legacy, err)
	}
	type Interface struct {
		Value fmt.Stringer   `skg:"value"`
		Items []fmt.Stringer `skg:"items"`
	}
	var interfaces Interface
	if err := Unmarshal([]byte("value: null items: [null]"), &interfaces); err != nil || interfaces.Value != nil || len(interfaces.Items) != 1 || interfaces.Items[0] != nil {
		t.Fatal(interfaces, err)
	}
}

func TestNativeNumerics(t *testing.T) {
	type F32 struct {
		Value float32 `skg:"value"`
	}
	type F64 struct {
		Value float64 `skg:"value"`
	}
	for _, src := range []string{"value: 0.1", "value: 16777217"} {
		_, err := DecodeSource[F32]([]byte(src), "", DecodeOptions{})
		expectNativeError(t, err, NativeInexactNumber, "/value")
		if _, err := DecodeSource[F32]([]byte(src), "", DecodeOptions{AllowLossyNumbers: true}); err != nil {
			t.Fatal(err)
		}
	}
	for _, src := range []string{"value: 0.5", "value: 16777218", "value: -9223372036854775808"} {
		if _, err := DecodeSource[F32]([]byte(src), "", DecodeOptions{}); err != nil {
			t.Fatal(err)
		}
	}
	_, err := DecodeSource[F64]([]byte("value: 9223372036854775807"), "", DecodeOptions{})
	expectNativeError(t, err, NativeInexactNumber, "/value")
	v, err := DecodeSource[F64]([]byte("value: -9223372036854775808"), "", DecodeOptions{})
	if err != nil || v.Value != math.MinInt64 {
		t.Fatalf("%#v %v", v, err)
	}
	_, err = DecodeSource[F32]([]byte("value: 10000000000000000000000000000000000000000.0"), "", DecodeOptions{AllowLossyNumbers: true})
	expectNativeError(t, err, NativeNumberOutOfRange, "/value")
	var legacy F32
	if err := Unmarshal([]byte("value: 0.1"), &legacy); err != nil || legacy.Value != 0.1 {
		t.Fatal(legacy, err)
	}
}

func TestNativeCollectionsAndBytes(t *testing.T) {
	type Record struct {
		ID   uint16  `skg:"id"`
		Note *string `skg:"note"`
	}
	type Config struct {
		Records []*Record         `skg:"records"`
		Matrix  [2][]int32        `skg:"matrix"`
		Headers map[string]*uint8 `skg:"headers"`
		Text    []byte            `skg:"text"`
		Dynamic Value             `skg:"dynamic"`
	}
	v, err := DecodeSource[Config]([]byte("records [null { id: 3 note: null }] matrix: [[1, 2], [3]] headers { \"a/b~c\": null b: 7 } text: \"hé😀\" dynamic: [[{ x: 1 }, null]]"), "", DecodeOptions{})
	if err != nil {
		t.Fatal(err)
	}
	if v.Records[0] != nil || v.Records[1].ID != 3 || v.Records[1].Note != nil || string(v.Text) != "hé😀" || *v.Headers["b"] != 7 || v.Dynamic.Type != TypeArray || v.Matrix[1][0] != 3 {
		t.Fatalf("%#v", v)
	}
	type Bytes struct {
		Text []byte `skg:"text"`
	}
	for _, src := range [][]byte{[]byte("text: [65, 255]")} {
		v, err := DecodeSource[Bytes](src, "", DecodeOptions{})
		if err != nil || !reflect.DeepEqual(v.Text, []byte{65, 255}) {
			t.Fatalf("%v %v", v, err)
		}
	}
	invalid := append(append([]byte("text: \""), 65, 255), '"')
	_, err = DecodeSource[Bytes](invalid, "invalid-utf8.skg", DecodeOptions{})
	d := expectNativeError(t, err, NativeParseError, "")
	var parse *ParseError
	if !errors.As(d, &parse) || parse.Diag.Code != CodeInvalidUTF8 || parse.Diag.Col != 9 {
		t.Fatal(err)
	}
	type Fixed struct {
		Items [2]uint8 `skg:"items"`
	}
	_, err = DecodeSource[Fixed]([]byte("items: [1]"), "", DecodeOptions{})
	expectNativeError(t, err, NativeTypeMismatch, "/items")
	encoded, err := Marshal(Fixed{Items: [2]uint8{2, 3}})
	if err != nil {
		t.Fatal(err)
	}
	got, err := DecodeSource[Fixed](encoded, "", DecodeOptions{})
	if err != nil || got.Items != [2]uint8{2, 3} {
		t.Fatal(got, err)
	}
}

type nativePort uint16

func (p *nativePort) DecodeSKG(ctx *DecodeContext, value Value) error {
	var n uint16
	if err := ctx.Decode(value, &n); err != nil {
		return err
	}
	*p = nativePort(n)
	return nil
}
func (p *nativePort) ValidateSKG(ctx *DecodeContext) error {
	if *p == 0 {
		return ctx.Fail(NativeCustomError, "port must be nonzero")
	}
	return nil
}

type nativeRecursive struct{}

func (p *nativeRecursive) DecodeSKG(ctx *DecodeContext, v Value) error { return ctx.Decode(v, p) }

var nativeHookCause = errors.New("application decoder failed")

type nativeFailing struct{ Values []int }

func (p *nativeFailing) DecodeSKG(ctx *DecodeContext, v Value) error {
	if len(p.Values) > 0 {
		p.Values[0] = 99
	}
	return nativeHookCause
}

func TestNativeHooks(t *testing.T) {
	type Config struct {
		Port nativePort `skg:"port"`
		Time time.Time  `skg:"time"`
	}
	v, err := DecodeSource[Config]([]byte("port: 8080 time: \"2026-01-02T03:04:05Z\""), "", DecodeOptions{})
	if err != nil || v.Port != 8080 || v.Time.Year() != 2026 {
		t.Fatal(v, err)
	}
	_, err = DecodeSource[Config]([]byte("port: 0"), "", DecodeOptions{})
	expectNativeError(t, err, NativeCustomError, "/port")
	_, err = DecodeSource[nativeRecursive](nil, "", DecodeOptions{})
	expectNativeError(t, err, NativeNestingTooDeep, "")
	type Failing struct {
		Value nativeFailing `skg:"value"`
	}
	defaults := Failing{Value: nativeFailing{Values: []int{7}}}
	err = DecodeSourceInto([]byte("value: 1"), "", &defaults, DecodeOptions{})
	expectNativeError(t, err, NativeCustomError, "/value")
	if !errors.Is(err, nativeHookCause) || defaults.Value.Values[0] != 7 {
		t.Fatal(defaults, err)
	}
}

type nullValidatedSlice []int

func (*nullValidatedSlice) ValidateSKG(ctx *DecodeContext) error {
	return ctx.Fail(NativeCustomError, "null container must not be validated")
}

func TestNativeValidatorSkipsNullTargets(t *testing.T) {
	type Config struct {
		Items nullValidatedSlice `skg:"items"`
	}
	value, err := DecodeSource[Config]([]byte("items: null"), "null.skg", DecodeOptions{})
	if err != nil || value.Items != nil {
		t.Fatalf("null target validation mismatch: %#v, %v", value, err)
	}
}

func TestNativeDefaultsTransaction(t *testing.T) {
	type Config struct {
		Count uint8           `skg:"count"`
		Items []int           `skg:"items"`
		Map   map[string]*int `skg:"map"`
	}
	n := 4
	defaults := Config{Count: 7, Items: []int{1}, Map: map[string]*int{"a": &n}}
	originalItems, originalMap := defaults.Items, defaults.Map
	err := DecodeSourceInto([]byte("items: [2] count: 256"), "", &defaults, DecodeOptions{AllowMissingFields: true})
	expectNativeError(t, err, NativeNumberOutOfRange, "/count")
	if defaults.Count != 7 || defaults.Items[0] != 1 || *defaults.Map["a"] != 4 {
		t.Fatal(defaults)
	}
	if err := DecodeSourceInto([]byte("@delete count"), "", &defaults, DecodeOptions{AllowMissingFields: true}); err != nil {
		t.Fatal(err)
	}
	defaults.Items[0], *defaults.Map["a"] = 9, 9
	if originalItems[0] != 1 || *originalMap["a"] != 4 {
		t.Fatal("defaults retain mutable caller aliases")
	}
	if defaults.Count != 7 {
		t.Fatal("delete must use native absence/default rule")
	}
	if err := DecodeSourceInto([]byte("map {}"), "", &defaults, DecodeOptions{AllowMissingFields: true}); err != nil || len(defaults.Map) != 0 {
		t.Fatal(defaults, err)
	}
	type Cycle struct {
		Next *Cycle `skg:"next"`
	}
	cycle := Cycle{}
	cycle.Next = &cycle
	err = DecodeSourceInto(nil, "", &cycle, DecodeOptions{AllowMissingFields: true})
	expectNativeError(t, err, NativeNestingTooDeep, "")
	type Unsafe struct{ Pointer unsafe.Pointer }
	var safe Unsafe
	if err := DecodeSourceInto(nil, "", &safe, DecodeOptions{}); err != nil {
		t.Fatal(err)
	}
	safe.Pointer = unsafe.Pointer(&n)
	err = DecodeSourceInto(nil, "", &safe, DecodeOptions{})
	expectNativeError(t, err, NativeUnsupportedType, "")
}

func TestNativeLocationsImportsAndParseErrors(t *testing.T) {
	type Config struct {
		Values []uint8 `skg:"a/b~c"`
	}
	_, err := DecodeSource[Config]([]byte("\"a/b~c\": [1, 256]"), "range.skg", DecodeOptions{})
	d := expectNativeError(t, err, NativeNumberOutOfRange, "/a~1b~0c/1")
	if d.Source != (SourceLocation{"range.skg", 1, 1}) {
		t.Fatal(d.Source)
	}
	dir := t.TempDir()
	for name, src := range map[string]string{"base.skg": "\"a/b~c\": [256]\nremoved: 1", "main.skg": "import \"base.skg\"\n@delete removed"} {
		if err := os.WriteFile(filepath.Join(dir, name), []byte(src), 0600); err != nil {
			t.Fatal(err)
		}
	}
	_, err = DecodeFile[Config](filepath.Join(dir, "main.skg"), DecodeOptions{RejectUnknownFields: true})
	d = expectNativeError(t, err, NativeNumberOutOfRange, "/a~1b~0c/0")
	if filepath.Base(d.Source.Path) != "base.skg" {
		t.Fatal(d.Source)
	}
	_, err = DecodeSource[Config]([]byte("x: ["), "broken.skg", DecodeOptions{})
	d = expectNativeError(t, err, NativeParseError, "")
	var parse *ParseError
	if !errors.As(err, &parse) || d.Source.Path != "broken.skg" {
		t.Fatal(err)
	}
	if _, err := DecodeSource[Config]([]byte("import \"does-not-exist.skg\""), "", DecodeOptions{}); err != nil {
		t.Fatal(err)
	}
}

func TestNativeAmbiguousMapping(t *testing.T) {
	type Duplicate struct {
		A int `skg:"a"`
		B int `skg:"a"`
	}
	_, err := DecodeSource[Duplicate](nil, "", DecodeOptions{})
	expectNativeError(t, err, NativeUnsupportedType, "")
	var legacy Duplicate
	if err := Unmarshal([]byte("a: 1"), &legacy); err != nil || legacy.A != 1 || legacy.B != 0 {
		t.Fatal(legacy, err)
	}
}

func TestNativeDynamicCycleAndChildContext(t *testing.T) {
	array := &Array{ElementType: TypeArray}
	value := Value{Type: TypeArray, Array: array}
	array.Items = []Value{value}
	ctx := DecodeContext{Source: SourceLocation{Path: "custom.skg"}}
	var out any
	err := ctx.Decode(value, &out)
	var d *DecodeError
	if !errors.As(err, &d) || d.Code != NativeNestingTooDeep {
		t.Fatal(err)
	}
	var n uint8
	err = ctx.DecodeChild(Value{Type: TypeInt, Int: 256}, "a/b~c", SourceLocation{"child.skg", 3, 5}, &n)
	d = expectNativeError(t, err, NativeNumberOutOfRange, "/a~1b~0c")
	if d.Source != (SourceLocation{"child.skg", 3, 5}) || ctx.FieldPath != "" {
		t.Fatal(d, ctx)
	}
}

func FuzzNativeDecode(f *testing.F) {
	for _, src := range []string{"", "count: 2 items: [1, null]", "count: null", "map { \"a/b\": 999 }", "items: [[{x:1}]]", "@delete count"} {
		f.Add(src)
	}
	type Config struct {
		Count uint8             `skg:"count"`
		Items []*int16          `skg:"items"`
		Map   map[string]*uint8 `skg:"map"`
	}
	f.Fuzz(func(t *testing.T, src string) {
		value, err := DecodeSource[Config]([]byte(src), "fuzz.skg", DecodeOptions{})
		if err != nil && !reflect.DeepEqual(value, Config{}) {
			t.Fatal("partial result")
		}
		defaults := Config{Count: 7, Items: []*int16{nil}}
		if err := DecodeSourceInto([]byte(src), "fuzz.skg", &defaults, DecodeOptions{}); err != nil {
			if defaults.Count != 7 || len(defaults.Items) != 1 || defaults.Items[0] != nil || defaults.Map != nil {
				t.Fatal("partial mutation")
			}
		}
	})
}
